### DISCUSSION TITLE
links

### DISCUSSION THREAD

**Discussion by derjavec** [denise-sandra/minishell#3] [General]
https://prod.liveshare.vsengsaas.visualstudio.com/join?5457E5D883C4047914F64F0D7D7AE6B3ECEC

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?574E779EE9401904F5D5AE5F83EA0E4BFD48

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?904FF81562DDCBE5F695DD422EF3463C1FD5

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?1D882415C65B21FEB2686D85F65B351DC403

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?1D882415C65B21FEB2686D85F65B351DC403

**Comment by SandraKanna** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?DE8FABE08C0643AB410A2B7C2BD1C19C8232

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?CE8DBBD8704245BACAB036E1B09E44BA62A7

**Reply by SandraKanna** [reply]
static	char	*erase_left_quotes(char *str, int q_type_l, int len)
{
	int		i;
	int		res;
	char	*new_str;
	
	i = len - 1;
	while (str[i] == q_type_l)
		i--;
	if ((len - 1 - i) % 2 == 0)
		res = 0;
	else
		res = 1;
	// printf("len: %d i: %d res: %d\n", len, i, res);
	new_str = malloc((i + 2 + res) * sizeof(char));
	if (new_str == NULL)
		return (NULL);
	ft_strlcpy(new_str, str, i + 2 + res);
	return (new_str);
}

static	char	*erase_right_quotes(char *str, int q_type_r, int len)
{
	int		i;
	int		res;
	char	*new_str;
	
	i = 0;
	while (str[i] == q_type_r)
		i++;
	if (i % 2 == 0)
		res = 0;
	else
		res = 1;
	new_str = malloc((len - i + res + 1) * sizeof(char));
	if (new_str == NULL)
		return (NULL);
	ft_strlcpy(new_str, str + i - res, len - i + res + 1);
	return (new_str);
}

static char	*copy_new_str(char *str, int q_type_r, int q_type_l, int len)
{
	char	*new_str;

	new_str = str;
	// printf("r: %d l: %d\n", q_type_r, q_type_l);
	if (q_type_r != 0)
	{
		new_str = erase_right_quotes(new_str, q_type_r, len);
		// printf("righ q %s\n", new_str);
	}
	if (q_type_l != 0)
	{
		new_str = erase_left_quotes(new_str, q_type_l, len);
		// printf("left q %s\n", new_str);
	}
	return (new_str);
}

char	*erase_extra_quotes(char *str, int len)
{
	char	*new_str;
	int	q_type_r;
	int	q_type_l;
	
	q_type_r = 0;
	q_type_l = 0;
	new_str = str;
	if (str[0] == 34 || str[0] == 39)
		q_type_r = str[0];
	if (str[len - 1] == 34 || str[len - 1] == 39)
		q_type_l = str[len - 1];
	if (q_type_r != 0 || q_type_l != 0)
		new_str = copy_new_str(str, q_type_r, q_type_l, len);
	if (new_str == NULL)
		return (NULL);
	if (ft_strncmp(str, new_str, len) == 0)
	{
		if (str != new_str)
			free (str);
		return (new_str);
	}	
	return (erase_extra_quotes(new_str, ft_strlen(new_str)));
}

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?8CB3746944B6C0C1FDA244F7633477B72CFA

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?C04255DD43044228061FF04C7908E16DE2CF

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?9467F588ACA1666C77A8277FCE64A63B05E8

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?E0DB9C8C1E0E71EDE71C1F108F66237756D6

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?969F63571A117F4B4FDC1EB7FB82061C0267

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?52F4C1E717F8E7042737814477B1120E78B8

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?E3F70FAE432BEC01E90C76E96AF3D675A1E2

**Comment by SandraKanna** [comment]
order_in:



void	order_in(t_mini *mini)
{
	t_token *cur = mini->token;
	t_token *prev = NULL;
	t_token *redir_head = NULL;
	t_token *redir_tail = NULL;
	t_token *cmd_head = NULL;
	t_token *cmd_tail = NULL;

	while (cur)
	{
		t_token *next = cur->next;

		if (cur->type == PIPE)
		{
			// Combine redirection tokens and command tokens for the current segment
			if (redir_tail)
			{
				redir_tail->next = cmd_head;
				cmd_tail->next = NULL; // End of command segment
				if (prev)
					prev->next = redir_head;
				else
					mini->token = redir_head;
				prev = cmd_tail ? cmd_tail : redir_tail;
			}
			else if (cmd_tail)
			{
				if (prev)
					prev->next = cmd_head;
				else
					mini->token = cmd_head;
				prev = cmd_tail;
			}

			// Set the pipe token
			if (prev)
				prev->next = cur;
			prev = cur;

			// Reset the lists for the next segment
			redir_head = redir_tail = NULL;
			cmd_head = cmd_tail = NULL;

			cur = next;
			continue;
		}
		else if (cur->type == IN || cur->type == HERE)
		{
			// Handle input redirection token
			if (redir_tail)
			{
				redir_tail->next = cur;
				redir_tail = cur;
			}
			else
			{
				redir_head = cur;
				redir_tail = cur;
			}

			// Check if the next token is a string (filename or heredoc delimiter)
			if (next && next->type == STRING)
			{
				redir_tail->next = next;
				redir_tail = next;
				cur = next->next;
			}
			else
			{
				cur = next;
			}
		}
		else
		{
			// Handle command/option/string token
			if (cmd_tail)
			{
				cmd_tail->next = cur;
				cmd_tail = cur;
			}
			else
			{
				cmd_head = cur;
				cmd_tail = cur;
			}
			cur->next = NULL;
			cur = next;
		}
	}

	// Combine the last segment's redirection tokens and command tokens
	if (redir_tail)
	{
		redir_tail->next = cmd_head;
		if (cmd_tail)
			cmd_tail->next = NULL;
		if (prev)
			prev->next = redir_head;
		else
			mini->token = redir_head;
	}
	else if (cmd_tail)
	{
		if (prev)
			prev->next = cmd_head;
		else
			mini->token = cmd_head;
	}
}

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?E510F854CA145D2591F5794026495B675634

**Comment by derjavec** [comment]
https://github.com/zstenger93/42_minishell_tester

**Comment by SandraKanna** [comment]
int exec_script(t_mini *mini, t_token *tmp) {
	char *name;
	int script;
	char *args[3];

	name = get_script_name(tmp);
	printf("1\n");
	init_args(args);
	if (name) {
		if (process_script(mini, name, &script) != 0)
			return 1;

		// Determina si es un script revisando si tiene un shebang.
		if (ft_strncmp(name + ft_strlen(name) - 3, ".sh", 3) == 0) {
			args[0] = get_shebang(mini, script, args[0]);
		} else {
			// Asume que es un ejecutable y usa el nombre del archivo.
			printf("4\n");
			args[0] = name;
		}

		if (!args[0])  // Asegura que args[0] no sea NULL
			return (close(script), 1);

		args[1] = name;
		args[2] = NULL;
		if (execve(args[0], args, mini->env_char) == -1) {
			printf("5 fail\n");
			mini->exit_status = 127;
			return (close(script), ft_error(mini, NULL, strerror(errno)), 1);
		}
		printf("5 ok\n");
		close(script);
	}
	return 0;
}

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?B3A984730C1E696AE9032F0FA4B7C7B52EB3

**Comment by derjavec** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?FBF48E6EC3D2CAC2758B2BB7CA8C174541CE

**Comment by SandraKanna** [comment]
https://prod.liveshare.vsengsaas.visualstudio.com/join?570F71603DFC4ABF99FE6EE48DCA04A15ED2

**Comment by SandraKanna** [comment]
static void	sigint_handler(int sig)
{
	(void)sig;
	g_sig = SIGINT;
	if (isatty(STDIN_FILENO))
	{
		ft_putstr_fd("\n", 1);
		rl_on_new_line();
		rl_replace_line("", 0);
		rl_redisplay();
	}
}

static void	sigquit_handler(int sig)
{
	(void)sig;
	g_sig = SIGQUIT;
	ft_putstr_fd("Quit\n", 1);
}

void	check_sigs(t_mini *mini)
{
	if (g_sig == SIGINT)
	{
		mini->exit_status = 130;
		// ft_putstr_fd("\n", 1);
		// rl_on_new_line();
		// rl_replace_line("", 0);
		// rl_redisplay();
		g_sig = 0;
	}
	else if (g_sig == SIGQUIT)
	{
		mini->exit_status = 131;
		g_sig = 0;
	}
}

void	enable_sigquit(void)
{
	struct sigaction	sa;

	sa.sa_handler = sigquit_handler;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;
	if (sigaction(SIGQUIT, &sa, NULL) == -1)
	{
		perror("sigaction for SIGQUIT");
		exit(1);
	}
}

void	init_handlers(void)
{
	struct sigaction	sa;

	sa.sa_handler = sigint_handler;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;
	if (sigaction(SIGINT, &sa, NULL) == -1)
	{
		perror("sigaction for SIGINT");
		exit(1);
	}
	sa.sa_handler = SIG_IGN;
	if (sigaction(SIGQUIT, &sa, NULL) == -1)
	{
		perror("sigaction for SIGQUIT");
		exit(1);
	}
}

**Comment by SandraKanna** [comment]
static void	minishell(t_mini *mini)
{
	char	*input;
	char	*prompt;

	while (!mini->should_exit)
	{
		init_handlers();
		g_sig = 0;
		mini->error = 0;
		prompt = get_dynamic_prompt();
		if (!prompt)
			break ;
		if (isatty(fileno(stdin)))
			input = readline(prompt);
		else
		{
			char *line;
			line = get_next_line(fileno(stdin));
			input = ft_strtrim(line, "\n");
			free(line);
		}
		free (prompt);
		if (!input)
			break ;
		enable_sigquit();
		check_sigs(mini);
		if (*input)
		{
			add_history(input);	
			parse_and_execute(mini, input);
		}
		clean_token_list(&(mini->token));
		clean_fd(mini);
		mini->token = NULL;
		free(input);
	}
}

**Comment by SandraKanna** [comment]
static char	**no_env(t_mini *mini)
{
	char	**envp;
	char	cwd[1024];

	envp = malloc(4 * sizeof(char *));
	if (envp == NULL)
		return (ft_error(mini, NULL, strerror(errno)), NULL);
	if (getcwd(cwd, 1024) == NULL)
		return (ft_error(mini, NULL, strerror(errno)), NULL);
	envp[0] = ft_strjoin("PWD=", cwd);
	if (!envp[0])
		return (free_tab(envp), ft_error(mini, NULL, strerror(errno)), NULL);
	envp[1] = ft_strdup("SHLVL=1");
	if (!envp[1])
		return (free_tab(envp), ft_error(mini, NULL, strerror(errno)), NULL);
	envp[2] = ft_strdup("PATH=/usr/bin/");
	if (!envp[2])
		return (free_tab(envp), ft_error(mini, NULL, strerror(errno)), NULL);
	envp[3] = NULL;
	return (envp);
}

**Comment by SandraKanna** [comment]
static t_mini	*init_minishell(char **envp)
{
	t_mini	*mini;

	mini = malloc(sizeof(t_mini));
	if (mini == NULL)
	{
		ft_error(mini, NULL, strerror(errno));
		exit(1);
	}
	ft_bzero(mini, sizeof(t_mini));
	if (envp[0] == NULL)
		envp = no_env(mini);
	mini->env = fill_env_struct(envp, mini);
	return (mini);
}

**Comment by derjavec** [comment]
if (tmp->cmd_tab && ft_strncmp(tmp->cmd_tab[0], "./minishell", longer_len(tmp->value, "./minishell")) == 0)
	{
		signal(SIGINT, SIG_IGN);
   		signal(SIGQUIT, SIG_IGN);
	}

**Comment by derjavec** [comment]
valgrind --suppressions=s.supp --track-origins=yes --show-leak-kinds=all --track-fds=yes --leak-check=full ./minishell

**Comment by SandraKanna** [comment]
https://enzodeg40.github.io/42-Evals/Cursus/Minishell/

**Comment by derjavec** [comment]
int	find_name(t_lst_env *lst, char *name)
{
	int	name_len;

	name_len = ft_strchr_int(name, '=');
	if (name_len == -1)
		name_len = ft_strlen(name);
	while (lst)
	{
		if (ft_strncmp(lst->name, name, name_len) == 0)
			return (1);
		lst = lst->next;
	}
	return (0);
}

**Comment by derjavec** [comment]
static void	add_exp(t_mini *mini, char *new_var, char *res)
{
	char		**split_new_exp;
	t_lst_env	*new_node;

	if (res == NULL)
	{
		split_new_exp = add_with_null_value(mini, new_var);
		if (!split_new_exp)
			return ;
	}
	else
	{
		if (find_name(mini->export, new_var) == 1)
			free_env_node(mini, mini->export, new_var);
		split_new_exp = split_env_vars(new_var, '=');
		if (split_new_exp == NULL)
			return (ft_error(mini, NULL, strerror(errno)));
	}
	new_node = ft_lstnew_env(split_new_exp[0], split_new_exp[1]);
	free_tab(split_new_exp);
	if (new_node == NULL)
		return (ft_error(mini, NULL, strerror(errno)));
	ft_lstadd_back_env(&mini->export, new_node);
}

**Comment by derjavec** [comment]
void	free_env_node(t_mini *mini, t_lst_env *env, char *name)
{
	t_lst_env	*temp;
	t_lst_env	*prev;
	t_lst_env	*next;
	size_t		len;

	temp = env;
	prev = NULL;
	next = NULL;
	len = ft_strchr_int(name, '=');
	if ((int)len == -1)
		len = ft_strlen(name);
	while (temp)
	{
		if (ft_strlen(temp->name) == len && \
			ft_strncmp(temp->name, name, len) == 0)
		{
			free_env_utils(mini, &temp, &prev, &next);
			return ;
		}
		prev = temp;
		temp = temp->next;
	}
}

**Comment by derjavec** [comment]
static void swap_nodes(t_lst_env *cur, t_lst_env *next)
{
	char	*name;
	char	*value;

	name = cur->name;
	value = cur->value;
	cur->name = next->name;
	cur->value = next->value;
	next->name = name;
	next->value = value;
}

static void	sort_list(t_lst_env *env)
{
	t_lst_env	*cur;
	t_lst_env	*next;
	int	s;

	s = 1;
	while (s != 0)
	{
		s = 0;
		cur = env;
		while (cur->next)
		{
			next = cur->next;
			if (ft_strncmp(cur->name, next->name, \
				longer_len(cur->name, next->name)) > 0)
			{
				swap_nodes(cur, next);
				s = 1;
			}
			cur = cur->next;
		}
	}

}

**Comment by derjavec** [comment]
static char	**add_with_null_value(t_mini *mini, char *new_var)
{
	int		n;
	char	**split;

	n = find_name(mini->export, new_var);
	if (n == 1)
		return (NULL);
	split = malloc(2 * sizeof(char *));
	if (!split)
		return (ft_error(mini, NULL, strerror(errno)), NULL);
	split[0] = ft_strdup(new_var);
	split[1] = NULL;
	if (!split[0])
		return (ft_error(mini, NULL, strerror(errno)), NULL);
	return (split);
}
